### ISSUE TITLE
Allow concurrent deploys

### ISSUE THREAD

**Issue by dorianmariecom** [basecamp/kamal#689]
It would be nice, I don't think we need a lock

**Comment by wenderjean** [comment]
@dorianmariecom Just for the sake of curiosity, do you have a real use case for concurrent deploys? I can just imagine deploying different environments at the same time

**Comment by dorianmariecom** [comment]
For instance I deploy a first fix and it's deployed while the second fix is building then deployed

**Comment by morgoth** [comment]
that could get messed up. Imagine having long migration in the first one. It could happen that the second one finishes before the first one, which wouldn't be something you expect.

I think it would be better for you to cancel the first deploy and deploy only the next one.

**Comment by dorianmariecom** [comment]
I tried that but then if I make too many commits only the last one will get deployed

**Comment by dorianmariecom** [comment]
What's the issue with releasing the lock while a deploy is in progress? It seems like it works fine

**Comment by djmb** [comment]
Kamal deployments are not designed to work concurrently. There are many things that could go wrong if you ran two at the same time. If it works ok, that's by luck and not design!

**Comment by igor-alexandrov** [comment]
@dorianmariecom I wrote about concurrent deploys in this article: https://igor.works/blog/evolution-of-github-action-for-kamal.

TL; DR; concurrency can be (and should be) achieved by using GitHub deployment groups with `cancel-in-progress: true`  and `if: ${{ cancelled() }}` options.

**Comment by igor-alexandrov** [comment]
@djmb seems like the issue can be closed.

**Comment by fredrikburmester** [comment]
One use case would be two environments like -d staging and -d production. It would be nice to be able deploy these at the same time.

**Comment by igor-alexandrov** [comment]
@fredrikburmester, what is the use case for the simultaneous deployment of two environments?

**Comment by fredrikburmester** [comment]
@igor-alexandrov I have a production and staging environment, they have completely separate containers but run on the same server. Sometimes I have a change that I want to push to both branches, but then it'll run both my develop and main git workflows, triggering `kamal deploy` for both production and develop.

**Comment by igor-alexandrov** [comment]
@fredrikburmester OK, now I understand. Currently, Kamal includes only the service name in the lock file name: https://github.com/basecamp/kamal/blob/main/lib/kamal/commands/lock.rb#L44. This allows you to deploy different services to the same host simultaneously.

We should also add the destination to the `lock_dir`.

@djmb do you agree?

**Comment by fredrikburmester** [comment]
> We should also add the destination to the lock_dir.

Yes, I agree.

@igor-alexandrov

**Comment by djmb** [comment]
> We should also add the destination to the lock_dir.

Yes that seems sensible 👍
